### PR TITLE
Add demo code

### PR DESCRIPTION
**Pull request by lorensr**

### CODE REVIEW

File: src/activities.ts [hunk 1]
```diff
@@ -1,3 +1,36 @@
-export async function greet(name: string): Promise<string> {
-  return `Hello, ${name}!`;
+import { Context } from '@temporalio/activity'
+import {
+  FulfillmentResult,
+  fulfillmentService,
+  inventoryService,
```
Comment (bergundy) [inline src/activities.ts:5]: You probably don't want to use global state in your app, better to instantiate these services on worker creation and inject into the activities.

File: src/api/order-v1.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import { decodeJWT } from '../auth'
+import { fulfillmentService, inventoryService, paymentService } from '../services'
+
+export default async (request: VercelRequest, response: VercelResponse) => {
+  const { itemId, quantity, addressId } = request.body
+  const { userId } = decodeJWT(request.headers.authorization)
+
+  await inventoryService.reserve({ itemId, quantity })
+  await paymentService.charge({ userId, itemId, quantity })
+  await fulfillmentService.sendPackage({ itemId, quantity, addressId })
```
Comment (bergundy) [inline src/api/order-v1.ts:15]: Not sure I follow, isn't this supposed to be in the workflow?

Comment (bergundy) [reply src/api/order-v1.ts:15]: Oh I got it

File: src/workflows.ts [hunk 3]
```diff
@@ -1,12 +1,25 @@
...
+    throw e
+  }
+
+  try {
+    await sendPackage(order)
+  } catch (e) {
+    await unreserve(order)
+    await refund(order)
+    throw e
+  }
```
Comment (bergundy) [inline src/workflows.ts]:
```suggestion
  try {
    await charge(order)
    try {
      await sendPackage(order)
    } catch (e) {
      await refund(order)
      throw e
    }
  } catch (e) {
    await unreserve(order)
    throw e
  }
```

Comment (bergundy) [reply src/workflows.ts]: Also, are we assuming that `refund` and `unreserve` will eventually succeed?

**Review by bergundy — APPROVED** [review]
Overall looks nice and definitely shows the strength of Temporal.
Didn't go into too much depth reviewing all the intermediate versions.

Maybe a bit too much detail and too many steps to follow.
I'd maybe try to reduce the number of steps.

File: src/api/order-v2.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  }
+
+  const fulfillment = await fulfillmentService.sendPackage({
+    itemId,
+    quantity,
+    addressId,
+  })
+  if (fulfillment.failed) {
+    await inventoryService.unreserve({ itemId, quantity })
+    await paymentService.refund({ userId, itemId, quantity })
```
Comment (bergundy) [inline src/api/order-v2.ts]: Should probably reverse these

File: src/activities.ts [hunk 5]
```diff
@@ -1,3 +1,36 @@
...
+export async function charge(order: Order): Promise<PaymentResult> {
+  return paymentService.charge({ ...order, requestId: getRequestId() })
+}
+export async function refund(order: Order): Promise<PaymentResult> {
+  return paymentService.refund({ ...order, requestId: getRequestId() })
+}
+
+export async function sendPackage(order: Order): Promise<FulfillmentResult> {
+  return fulfillmentService.sendPackage({ ...order, requestId: getRequestId() })
 }
```
Comment (bergundy) [inline src/activities.ts:36]: These APIs should have an idempotency key

File: src/activities.ts [hunk 6]
```diff
@@ -1,3 +1,36 @@
...
+export async function unreserve(order: Order): Promise<ReservationResult> {
+  return inventoryService.unreserve({ ...order, requestId: getRequestId() })
+}
+
+export async function charge(order: Order): Promise<PaymentResult> {
+  return paymentService.charge({ ...order, requestId: getRequestId() })
+}
+export async function refund(order: Order): Promise<PaymentResult> {
+  return paymentService.refund({ ...order, requestId: getRequestId() })
+}
```
Comment (bergundy) [inline src/activities.ts:26]: Charge and refund should probably not know anything about inventory, just amount to charge

File: src/api/order-v3.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  const fulfillment = await retry(() =>
+    fulfillmentService.sendPackage({
+      itemId,
+      quantity,
+      addressId,
+    })
+  )
+  if (fulfillment.failed) {
```
Comment (bergundy) [inline src/api/order-v3.ts:69]: I think retry throws, you should catch the error and do the rollback

File: src/api/order-v4.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    } catch (e) {
+      error = e
+      await new Promise((resolve) => setTimeout(resolve, initialInterval * Math.pow(2, attempt++)))
+    }
+  }
+  if (error) {
+    throw error
+  }
+  if (!result) {
+    throw new Error('Reached max attempts')
```
Comment (bergundy) [inline src/api/order-v4.ts]: Typically you'd also want to rethrow the last error

File: src/api/order-v4.ts [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        new Promise<undefined>((resolve) => setTimeout(resolve, callTimeout, undefined)),
+      ])
+      const timedOut = result === undefined
+      if (timedOut) {
+        throw new Error('Timed out')
+      }
+      break
+    } catch (e) {
+      error = e
+      await new Promise((resolve) => setTimeout(resolve, initialInterval * Math.pow(2, attempt++)))
```
Comment (bergundy) [inline src/api/order-v4.ts]: You could use https://nodejs.org/api/timers.html#timerspromisessettimeoutdelay-value-options if assuming node >= 16

File: src/api/order-v5.ts [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        new Promise<undefined>((resolve) => setTimeout(resolve, callTimeout, undefined)),
+      ])
+      const timedOut = result === undefined
+      if (timedOut) {
+        throw new Error('Timed out')
+      }
+      break
+    } catch (e) {
+      error = e
+      await new Promise((resolve) => setTimeout(resolve, initialInterval * Math.pow(2, attempt++)))
```
Comment (bergundy) [inline src/api/order-v5.ts]: This is hard to follow, just put increment the attempt in the loop:

```
for (let attempt = 1; attempt <= maxAttempts; ++attempt);
```

File: src/api/order-v6.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    error = e
+  })
+  if (!reservation || reservation.failed) {
+    response.status(400).send(reservation ? `Don't have enough inventory` : error.message)
+    return
+  }
+
+  const payment = await retry(() => paymentService.charge({ userId, itemId, quantity, requestId })).catch((e) => {
+    error = e
+  })
```
Comment (bergundy) [inline src/api/order-v6.ts]: Why not just throw errors here?

File: src/api/order-v7.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+  if (error) {
+    throw error
+  }
+  if (!result) {
+    throw new Error('Reached max attempts')
+  }
+  return result
+}
+
+enum STATES {
```
Comment (bergundy) [inline src/api/order-v7.ts]:
```suggestion
enum OrderState {
```

File: src/activities.ts (hunk 6 above)
Comment (lorensr) [reply src/activities.ts:26]: Yeah, for simplicity's sake I think I'll gloss over / hide amount calculation

File: src/api/order-v7.ts [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    const result = await orders.insertOne({
+      itemId,
+      quantity,
+      addressId,
+      requestId,
+      userId,
+      state: STATES.CREATED,
+    })
+    if (!result.acknowledged) {
+      response.status(500).send(`Unable to write to database`)
```
Comment (bergundy) [inline src/api/order-v7.ts]: Don't reveal there are database issues

```suggestion
      response.status(500).send('Failed to initiate order')
```

File: src/api/order-v3.ts (hunk 7 above)
Comment (lorensr) [reply src/api/order-v3.ts:69]: Yep that's part of the narrative, added in subsequent step

File: src/api/order-v7.ts [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+      fulfillmentService.sendPackage({
+        itemId,
+        quantity,
+        addressId,
+        requestId,
+      })
+    ).catch((e) => {
+      error = e
+    })
+    if (!fulfillment || fulfillment.failed) {
```
Comment (bergundy) [inline src/api/order-v7.ts:129]: This just makes everything hard to follow, just throw errors if there are errors

File: src/api/order-v7.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+  PAID,
+  FAILED_TO_CHARGE,
+  FAILED_TO_CHARGE_UNRESERVED,
+  FULFILLED,
+  FAILED_TO_FULFILL,
+  FAILED_TO_FULFILL_UNRESERVED,
+  FAILED_TO_FULFILL_REFUNDED,
+}
+
+export default async (request: VercelRequest, response: VercelResponse) => {
```
Comment (bergundy) [inline src/api/order-v7.ts:57]: Assuming you'll talk about how this isn't retryable

File: src/api/order-v5.ts (hunk 10 above)
Comment (lorensr) [reply src/api/order-v5.ts]: Yeah, thanks! I blame Copilot 😁

File: src/api/order-v6.ts (hunk 11 above)
Comment (lorensr) [reply src/api/order-v6.ts]: You mean try/catch block instead of `.catch()`? Fewer lines/indentation.

File: src/api/order-v7.ts (hunk 15 above)
Comment (lorensr) [reply src/api/order-v7.ts:57]: Like would need worker pool to check for non-terminal-status orders that haven't been updated in a while? Yep
